### PR TITLE
DAOS-8722 dfuse: Document and log readdir kernel support.

### PR DESCRIPTION
**Pull request by ashleypittman**
Set the flag to allow the kernel to cache readdir requests,
update the documentation and version/log output to make it
clear if this feature is compiled in or not.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: docs/user/filesystem.md [hunk 1]
```diff
@@ -314,6 +314,11 @@ files will use the page cache.  This default value for this is "off".
 With no options specified attr and dentry timeouts will be 1 second, dentry-dir
 and ndentry timeouts will be 5 seconds, and data caching will be enabled.
 
+Readir caching will be enabled when the dfuse-dentry-time setting is non-zero and when supported by
+libfuse however on many distributions the system libfuse is not able to support this feature.
+Libfuse version 3.5.0 is required at both compile and run-time.  Use `dfuse --version` or the
+runtime logs to see the fuse version used and if the feature is compiled into dfuse.
```
Comment (mchaarawi) [inline docs/user/filesystem.md:320]: would be helpful to add the line that corresponds in the runtime logs to look for

File: src/client/dfuse/ops/opendir.c [hunk 2]
```diff
@@ -31,8 +31,21 @@ dfuse_cb_opendir(fuse_req_t req, struct dfuse_inode_entry *ie, struct fuse_file_
...
 
 #if HAVE_CACHE_READDIR
-	if (ie->ie_dfs->dfc_dentry_timeout > 0)
+	/* TODO: Keep track of age of cache and selectively enable this bit.
+	 * The kernel keeps the readdir data in the page cache which has no time based
+	 * expiry mechanism, it would be better to keep track of when the last successful
+	 * opendir/readir/closedir calls (excudling where seekdir was used) happened
+	 * and set the keep_cache bit here based on how old the data in cache is.
+	 * fuse can peek into the page cache so we could identify if there is data in cache
+	 * or not but it's not possible to know which calls popultaed it so getting
```
Comment (jamesanunez) [inline src/client/dfuse/ops/opendir.c:40]: "populted" should be "populated"

File: docs/user/filesystem.md (hunk 1 above)
Comment (mchaarawi) [reply docs/user/filesystem.md:320]: oh or maybe it's just that new info log you added.
i recall there was a parameter printed that i had to check

File: src/client/dfuse/ops/opendir.c [hunk 3]
```diff
@@ -31,8 +31,21 @@ dfuse_cb_opendir(fuse_req_t req, struct dfuse_inode_entry *ie, struct fuse_file_
...
+	 * and set the keep_cache bit here based on how old the data in cache is.
+	 * fuse can peek into the page cache so we could identify if there is data in cache
+	 * or not but it's not possible to know which calls populated it so getting
+	 * absolute timings is likely impossible but if dfuse assumes worst-case in terms
+	 * of age this will mean that the cache timeouts are slightly shorter than
+	 * configured which is unlikely to be a problem.
+	 */
+	if (ie->ie_dfs->dfc_dentry_timeout > 0) {
 		fi_out.cache_readdir = 1;
+		fi_out.keep_cache    = 1;
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/ops/opendir.c:47]: I think the header file comment is wrong.  We already set it in open which is for files, the real change in this PR is to also set it in opendir which should enable the feature for directories as well.  I think they've updated the flag to have a second meaning and not updated that comment.

I can try this locally however, and will report back.

File: docs/user/filesystem.md [hunk 4]
```diff
@@ -314,6 +314,11 @@ files will use the page cache.  This default value for this is "off".
 With no options specified attr and dentry timeouts will be 1 second, dentry-dir
 and ndentry timeouts will be 5 seconds, and data caching will be enabled.
 
+Readir caching will be enabled when the dfuse-dentry-time setting is non-zero and when supported by
+libfuse however on many distributions the system libfuse is not able to support this feature.
```
Comment (daltonbohning) [inline docs/user/filesystem.md:318]: Grammar nitpick
```suggestion
libfuse; however, on many distributions the system libfuse is not able to support this feature.
```

**Review by daltonbohning — COMMENTED** [review]
Is there potential for this to break any existing test cases?

File: src/client/dfuse/ops/opendir.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/ops/opendir.c:47]: No, you're entirely right this setting is not required.

File: docs/user/filesystem.md (hunk 4 above)
Comment (ashleypittman) [reply docs/user/filesystem.md]: Done.

**Comment by ashleypittman** [comment]
So on reflection and testing there's no logic changes needed here, what this PR does is document and current behaviour and add logging to allow users to know if it's working.  In light of that it would be worth having the test team run benchmarks with master (with our without this PR) but a later libfuse to see if there's an improvement or not or if the hack in https://github.com/daos-stack/daos/pull/6776 is helping as well.

https://github.com/daos-stack/daos/pull/9370 is a PR I created with the cache configuration changes that came out of this work.

**Review by jamesanunez — APPROVED** [review]

**Pull request merged**